frontend/src/modules/placement/classes.ts

```diff
@@ -0,0 +1,874 @@
+        }
+
+        const [axisRow = [''], axisCol = ['']] = this.spec
+        const placementLocations = this.placementDestinations(samples, `${axisRow[0]}${axisCol[0]}`, PlacementType.PATTERN)
+        if (placementLocations.length !== samples.length) {
+            throw new Error(`Placement locations length does not match samples length`)
+        }
+        for (let i = 0; i < placementLocations.length; i++) {
+            const sample = samples[i]
+            const destinationCell = placementLocations[i]
+            sample.placeAtCell(destinationCell.rawIdentifier())
+        }
+    }
+    undoPlacements() {
+        let placements = this.getPlacements(true).filter((s) => !s.sample.fromCell?.fromContainer?.sameContainerAs(this))
+        if (placements.length === 0) {
+            placements = this.getPlacements(false).filter((s) => !s.sample.fromCell?.fromContainer?.sameContainerAs(this))
+        }
+        for (const selection of placements) {
+            selection.cell.unplaceSample(selection.sample)
+        }
+    }
+
+    isPlacementSelected(placedSample: SamplePlacementIdentifier) {
+        return this.getCell(placedSample.cell).isSampleSelected(placedSample.sample)
+    }
+    selectPlacement(placedSample: SamplePlacementIdentifier) {
+        this.getCell(placedSample.cell).setSelection(placedSample.sample, true)
+    }
+    unSelectPlacement(placedSample: SamplePlacementIdentifier) {
+        this.getCell(placedSample.cell).setSelection(placedSample.sample, false)
+    }
+    togglePlacementSelections(...entries: SamplePlacementIdentifier[]) {
+        const allSelected = entries.every(s => this.isPlacementSelected(s))
+        if (allSelected) {
+            for (const entry of entries) {
+                this.unSelectPlacement(entry)
+            }
+        } else {
+            for (const entry of entries) {
+                this.selectPlacement(entry)
+            }
+        }
+    }
+
+    clearPreviews() {
+        for (const cellKey in this.state.cells) {
+            this.state.cells[cellKey].preview = null
+        }
+    }
+    setPreviews(sources: Pick<CellIdentifier, 'coordinates'>[], destinations: CellIdentifier[]) {
+        this.clearPreviews()
+        for (let index = 0; index < sources.length && index < destinations.length; index++) {
+            const source = sources[index]
+            const destination = destinations[index]
+            const cell = this.getCell(destination)
+            cell.state.preview = source.coordinates ?? ''
+            if (cell.getSamples(true).length > 0) {
+                this.placementState.error ??= `Sample ${cell.getSamples(true)[0].name} already exists in cell ${cell.coordinates}`
+            }
+        }
+    }
+
+    compareSamples(a: SamplePlacementIdentifier, b: SamplePlacementIdentifier) {
+        const MAX = 128
+
+        const A = this.#resolveSamplePlacementIdentifier(a)
```

Comment:
This here appear very inefficient ... you are allocating objects in a sorting loop this can likely be done better with passing the objects from outside the sorting call. A single item will be reallocated multiple time here. You also incur the risk of modifying your item between calls.